core/commonMain/src/kotlinx/serialization/internal/ObjectSerializer.kt

```diff
@@ -17,6 +17,15 @@ import kotlinx.serialization.encoding.*
 @PublishedApi
 @OptIn(ExperimentalSerializationApi::class)
 internal class ObjectSerializer<T : Any>(serialName: String, private val objectInstance: T) : KSerializer<T> {
+    @PublishedApi
+    internal constructor(
+        serialName: String,
+        objectInstance: T,
+        classAnnotations: Array<Annotation>
+    ) : this(serialName, objectInstance) {
+        (this.descriptor as SerialDescriptorImpl).annotations = classAnnotations.asList()
```

Comment:
The problem is, the secondary constructor is called after all `init` blocks and properties initializers. I considered the following alternatives:
- Mutable (to some reasonable extent) serialDescriptorImpl — current solution. I can add `buildSerialDescriptorInternal` function to avoid downcast to be sure this won't break.
- Store the builder in property, call `.build` on every access to `public val descriptor` — I don't like copying every time
- Store descriptor in private mutable property, rebuild it from scratch (because we do not have a copying mechanism) every time. Seems reasonable, although a bit boilerplaite-ish. However, I do not see any strong objections to this solution, so I can rewrite everything to it